src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -958,34 +957,97 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         winrt::hstring currentCommandline,
         winrt::hstring currentWorkingDirectory)
     {
+        // enumerate all the parent directories we want to import snippets from
+        std::filesystem::path directory{ std::wstring_view{ currentWorkingDirectory } };
+        std::vector<std::filesystem::path> directories;
+        while (!directory.empty())
         {
-            // Check if there are any cached commands in this directory.
+            directories.push_back(directory);
+            const auto parentPath = directory.parent_path();
+            if (directory == parentPath)
+            {
+                break;
+            }
+            directory = parentPath;
+        }
+
+        {
+            // Check if all the directories are already in the cache
             const auto& cache{ _cwdLocalSnippetsCache.lock_shared() };
 
-            const auto cacheIterator = cache->find(currentWorkingDirectory);
-            if (cacheIterator != cache->end())
+            bool allSnippetsCached = true;
+            for (const auto& dir : directories)
+            {
+                if (!cache->contains(dir))
+                {
+                    allSnippetsCached = false;
+                    break;
+                }
+            }
+            if (allSnippetsCached)
             {
-                // We found something in the cache! return it.
+                // Load snippets from directories in reverse order.
+                // This ensures that we prioritize snippets closer to the cwd.
+                // The map makes it easy to avoid duplicates.
+                std::unordered_map<hstring, Model::Command> localSnippetsMap;
+                for (decltype(directories)::reverse_iterator rit = directories.rbegin(); rit != directories.rend(); ++rit)
+                {
+                    // register snippets from cache
+                    std::ranges::for_each(cache->at(*rit), [&localSnippetsMap](const auto& kvPair) {
+                        localSnippetsMap.insert_or_assign(kvPair.first, kvPair.second);
+                    });

```

Comment:
A plain `for(auto&& kvPair : cache->at(*rit))` seems more readable to me.
This and the previous comment are repeated once more below.